### PR TITLE
Add (and fix) some additional tests for updates

### PR DESCRIPTION
**Pull request by acron0**
This includes metadata and sharing

### CODE REVIEW

File: src/witan/httpapi/components/activities.clj [hunk 1]
```diff
@@ -160,17 +160,30 @@
                           {:partition-key file-id})
     (return-receipt id)))
 
+(defn- conform-metadata-updates
+  "This fn attempts to heal any differences between specs across the applications"
+  [m]
+  (log/debug "Conforming metadata updates:" m)
+  (let [m' (reduce-kv (fn [a k v]
+                        (assoc a k (s/unform k (s/conform k v)))) {} m)]
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj:168]: what is `unform` doing here? And worthwhile calling k spec?

File: src/witan/httpapi/components/activities.clj [hunk 2]
```diff
@@ -160,17 +160,30 @@
...
-    (create-receipt! database user id)
-    (send-valid-command!* comms (merge {::command/id id
-                                        ::command/type :kixi.datastore.metadatastore/update
-                                        ::command/version "1.0.0"
-                                        :kixi/user user}
-                                       (assoc metadata-updates :kixi.datastore.metadatastore/id file-id))
-                          {:partition-key file-id})
-    (return-receipt id)))
+  (if (empty? metadata-updates)
+    [BAD_REQUEST {:error "Update can't be empty"} nil]
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj:175]: I like the `BAD_REQUEST` constant. This saved me a mental hiccup. thx

File: src/witan/httpapi/components/activities.clj [hunk 3]
```diff
@@ -160,17 +160,30 @@
...
+    (let [id (comms/uuid)]
+      (create-receipt! database user id)
+      (send-valid-command!* comms (merge {::command/id id
+                                          ::command/type :kixi.datastore.metadatastore/update
+                                          ::command/version "1.0.0"
+                                          :kixi/user user}
+                                         (-> metadata-updates
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj]: this bit in the `->` feels like something I'd extract into a function and have a unit or generative/spec based test for.

File: src/witan/httpapi/components/activities.clj [hunk 4]
```diff
@@ -262,11 +275,23 @@
         (complete-receipt! db command-id nil)))))
 
 (defmethod on-event
+  [:kixi.datastore.metadatastore/update-rejected "1.0.0"]
+  [db {:keys [kixi.comms.event/payload] :as event}]
+  (let [command-id (:kixi.comms.command/id event)]
+    (when-let [receipt (retreive-receipt db command-id)]
+      (let [file-id (or (get-in payload [::ms/id])
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj]: again, a function and unit/spec test for this `or`

File: src/witan/httpapi/components/activities.clj [hunk 5]
```diff
@@ -262,11 +275,23 @@
...
+        (create-error! db command-id file-id (-> payload :reason name))
+        (complete-receipt! db command-id (str "/api/files/" file-id "/errors/" command-id)))))  )
+
+(defmethod on-event
   [:kixi.datastore.metadatastore/sharing-change-rejected "1.0.0"]
   [db {:keys [kixi.comms.event/payload] :as event}]
   (let [command-id (:kixi.comms.command/id event)]
     (when-let [receipt (retreive-receipt db command-id)]
-      (let [file-id (get-in payload [:original :kixi.datastore.metadatastore/id])]
+      (let [file-id (or (get-in payload [:kixi.datastore.metadatastore/id])
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj]: again function + unit/spec test

**Review by otfrom — CHANGES_REQUESTED** [review]
I'd like to start seeing the data construction/translation bits happening in separate functions that could be unit or spec tested. The feel a bit tied to things that do side effecting things like comms/http and require slow integration tests.

File: src/witan/httpapi/components/activities.clj (hunk 1 above)
Comment (acron0) [reply src/witan/httpapi/components/activities.clj:168]: `conform` then `unform` is a cheeky coercion.

File: src/witan/httpapi/components/activities.clj [hunk 6]
```diff
@@ -160,17 +160,31 @@
...
-                                        ::command/type :kixi.datastore.metadatastore/update
-                                        ::command/version "1.0.0"
-                                        :kixi/user user}
-                                       (assoc metadata-updates :kixi.datastore.metadatastore/id file-id))
-                          {:partition-key file-id})
-    (return-receipt id)))
+  (if (empty? metadata-updates)
+    [BAD_REQUEST {:error "Update can't be empty"} nil]
+    (let [id (comms/uuid)
+          update-payload (-> metadata-updates
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj:177]: update-payload still isn't unit/spec testable, but at least it is clearer what to refactor

File: src/witan/httpapi/components/activities.clj [hunk 7]
```diff
@@ -262,13 +285,28 @@
...
     (when-let [receipt (retreive-receipt db command-id)]
-      (let [file-id (get-in payload [:original :kixi.datastore.metadatastore/id])]
-        (create-error! db command-id file-id (-> payload :reason name))
-        (complete-receipt! db command-id (str "/api/files/" file-id "/errors/" command-id)))))  )
+      (if-let [file-id (get-metadatastore-id payload)]
+        (complete-receipt!
+         db
+         command-id
+         (create-error! db command-id file-id (-> payload :reason name)))
+        (log/error "Could't find a metadatastore ID in payload:" payload))))
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj:297]: should this throw an exception?

File: src/witan/httpapi/components/activities.clj [hunk 8]
```diff
@@ -262,13 +285,28 @@
...
+    [:kixi.datastore.metadatastore/sharing-change-rejected "1.0.0"]
+    [db {:keys [kixi.comms.event/payload] :as event}]
+    (let [command-id (:kixi.comms.command/id event)]
+      (when-let [receipt (retreive-receipt db command-id)]
+        (if-let [file-id (get-metadatastore-id payload)]
+          (complete-receipt!
+           db
+           command-id
+           (create-error! db command-id file-id (-> payload :reason name)))
+          (log/error "Could't find a metadatastore ID in payload:" payload))))))
```
Comment (otfrom) [inline src/witan/httpapi/components/activities.clj:309]: should this throw an exception? (or should the exception happen in `get-metadatastore-id` as there should be an ID by this point.

**Review by otfrom — APPROVED** [review]
still some niggles, but we can move on unless you want me to be picky

**Pull request merged**